Add a new parameter convention @in_cxx

ahatanaka opened the pull request.

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

Sadewoabdi commented:
How can I help

Later, I'll take a look at the files that have been repaired and see if
there are still red ones. If there are still red ones, I'll let you know,
thank you.

Pada Min, 21 Apr 2024 03.00, Akira Hatanaka ***@***.***>
menulis:

> @swift-ci <https://github.com/swift-ci> please test
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/73164#issuecomment-2067756182>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BH2MAXD7CZ763IVDODI4YT3Y6K3LXAVCNFSM6AAAAABGQQ337KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRXG42TMMJYGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

Sadewoabdi commented:
include/swift/SIL/SILDeclRef.h Try to see if there's anything wrong

Pada Min, 21 Apr 2024 15.24, Akira Hatanaka ***@***.***>
menulis:

> @swift-ci <https://github.com/swift-ci> please test
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/73164#issuecomment-2067938408>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BH2MAXABCBLRWTYURDNE6C3Y6NSRPAVCNFSM6AAAAABGQQ337KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRXHEZTQNBQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci please test

ahatanaka committed:
Add a new parameter convention `@in_cxx` for non-trivial C++ classes
that are passed indirectly and destructed by the caller

Fix a bug where `@in`, which indicates the callee is responsible for
destroying the passed object, was being used to pass such classes.

rdar://122707697

ahatanaka commented:
@swift-ci please test

ahatanaka committed:
Address feedback from @atrick and @nate-chandler

- @in_cxx is handled the same way as @in in callers and @in_guaranteed
  in callees (except for mutation). Fix checks in various places
  accordingly.

- Emit copies of @in_cxx parameters and destroy them in thunks.

ahatanaka committed:
Add type attribute for @in_cxx

ahatanaka committed:
Emit destroy_addr when inlining a function that has @in_cxx parameters

ahatanaka committed:
Fix mangling of @in_cxx

'C' is already taken by @convention(c).

ahatanaka committed:
Add missing case

ahatanaka committed:
Fix function names in tests after mangling change

ahatanaka committed:
Fix tests

ahatanaka committed:
Fix check strings

ahatanaka committed:
Create two versions (for caller and callee) of the functions that answer
questions about parameter convention

Create two versions of the following functions:

isConsumedParameter
isGuaranteedParameter
SILParameterInfo::isConsumed
SILParameterInfo::isGuaranteed
SILArgumentConvention::isOwnedConvention
SILArgumentConvention::isGuaranteedConvention

These changes will be needed when we add a new convention for
non-trivial C++ types as the functions will return different answers
depending on whether they are called for the caller or the callee. This
commit doesn't change any functionality.

ahatanaka committed:
Fix build error

ahatanaka committed:
Fix build error

ahatanaka committed:
Fix isConsumedParameter, etc.

ahatanaka commented:
@swift-ci please test
